add is-write-index flag to aliases

talevy opened the pull request:
This commit adds the is-write-index flag for aliases.
It allows requests to set the flag, and responses to display the flag.
It does not validate and/or affect any indexing/getting/updating behavior
of Elasticsearch -- this will be done in a follow-up PR.

This is a follow-up to: https://github.com/elastic/elasticsearch/pull/30703
It made more sense to split the above PR into two PRs. One for introducing the `is_write_index`
flag, and another to use it for validation and index resolution.

elasticmachine commented:
Pinging @elastic/es-core-infra

bleskes commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.put_alias/all_path_options.yml:
```diff
@@ -27,9 +27,9 @@ setup:
         name: alias
 
-  - match: {test_index1.aliases.alias: {}}
+  - match: {test_index1.aliases.alias: { is_write_index: false }}
```
why  are these false? this is the only index in this alias so it should be true?

bleskes commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/Alias.java:
```diff
@@ -60,6 +62,8 @@
     @Nullable
     private String searchRouting;
 
+    private boolean writeIndex;
```
How do we specify "default"? don't we want a Boolean here?

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -516,7 +516,7 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                 for (Alias alias : request.aliases()) {
                     AliasMetaData aliasMetaData = AliasMetaData.builder(alias.name()).filter(alias.filter())
-                        .indexRouting(alias.indexRouting()).searchRouting(alias.searchRouting()).build();
+                        .indexRouting(alias.indexRouting()).searchRouting(alias.searchRouting()).writeIndex(alias.isWriteIndex()).build();
```
we need to resolve the is write index here (for when it's not specified).

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/AliasAction.java:
```diff
@@ -112,7 +121,8 @@ boolean removeIndex() {
             aliasValidator.validate(alias, indexRouting, filter);
             AliasMetaData newAliasMd = AliasMetaData.newAliasMetaDataBuilder(alias).filter(filter).indexRouting(indexRouting)
-                    .searchRouting(searchRouting).build();
+                    .searchRouting(searchRouting).writeIndex(writeIndex).build();
```
what happens if writeIndex is null?

bleskes commented on docs/reference/indices/aliases.asciidoc:
```diff
@@ -244,6 +244,65 @@ GET /alias2/_search?q=user:kimchy&routing=2,3
+==== Write Index
+
+It is possible to associate the index pointed to by an alias as the write index.
+When specified, all index and update requests against an alias that point to multiple
```
I think that if you explicitly set a the write flag to false, we shouldn't write even if it's the only index the alias points to.

talevy commented:
@bleskes thanks for the initial review. As I mentioned outside of Github, I believe we were misaligned in terms of how to break apart the original PR that both added the flag, and leveraged it to resolve index name expressions.

I've updated this PR to re-add validation and auto-upgrading of the `is_write_index`. This means that alias-index configurations with `is_write_index=null` in both alias-add-actions and index-creations will be set to `true` when there is no existing write index for that alias. In both cases that a promotion to `true` is not possible, `false` will be set.

Although AliasMetaData stores this value as a `Boolean`, any cluster-state changes that occur via index-creation or alias-updates will resolve this value to either `true` or `false`. AliasMetaData#writeIndex must stay a Boolean because of its use in templates. Templates need not define this field explicitly, and no smart unboxing should occur.

I've also added some unit tests that should account for these scenarios. (still iterating to catch all the CI tests that need updating).

talevy commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.put_alias/all_path_options.yml:
```diff
@@ -45,9 +45,9 @@ setup:
+  - match: {foo.aliases.alias: { is_write_index: false }}
 
 ---
 "put alias in * index":
```
@bleskes, we discussed early about how it felt awkward to allow the defaulting of `is_write_index -> true` when 
not specified in alias updates apis (in addition to index creation). Towards the end of our discussion it seemed 
reasonable to rethink this and maybe enable this behavior. Although ordering was clear in the index-aliases api where actions are specified in a deterministic order, `indices.put_alias` allows resolution to multiple indices in such a way that ordering is undefined.

For these reasons, do you think it would be OK if we only default (when possible) `is_write_index -> true` during index creation?

talevy commented:
@bleskes I've updated to reflect the changes we mentioned offline

- move all validation logic to run at the time that `MetaData` is built
- default `is_write_index (null)` to `true` if-and-only-if there are no other aliases, `false` otherwise

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/AliasOrIndex.java:
```diff
@@ -134,8 +149,11 @@ public AliasMetaData getFirstAliasMetaData() {
         }
 
-        void addIndex(IndexMetaData indexMetaData) {
+        void addIndex(IndexMetaData indexMetaData, @Nullable Boolean isWriteIndex) {
```
Instead of doing this like this maybe only do it at the end? add a method called "computeAndValidateWriteIndex" and let that one do two things:
- scan the all index metadata and extracts the writeIndex, if any
- validate we have no two indices which are marked with the index_write flags.

you can use org.apache.lucene.util.SetOnce to store the writeIndex reference. Then you can assert that once `computeAndValidateWriteIndex` is called you can't call addIndex anymore.

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/AliasValidator.java:
```diff
@@ -145,6 +148,29 @@ public void validateAliasFilter(String alias, byte[] filter, QueryShardContext q
+     * @param aliasAndIndexLookup The current state of aliases and indices in the cluster state
+     * @throws IllegalStateException if any alias has more than one write index
+     */
+    public static void validateAliasWriteOnly(Map<String, AliasOrIndex> aliasAndIndexLookup) {
```
see my other comment.

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -701,7 +702,8 @@ private static IndexNotFoundException indexNotFoundException(String expression)
                             .filter(e -> e.getValue().isAlias() == false)
-                            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+                            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
+                                (v1, v2) -> {throw new IllegalStateException("no duplicate indices should exist");}, TreeMap::new));
```
out of curiosity - why did you add this?

talevy replied:
oh woops. this was not meant to stay. I added this to debug where the ordering of indices was being lost. will revert

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -557,6 +557,25 @@ public IndexMetaData getIndexSafe(Index index) {
+     * @param alias the alias to search for a write-index with
+     * @return the write-index {@link Index} for <code>alias</code>, null if no write-index is configured
+     */
+    public Index getWriteIndex(String alias) {
```
this is usage? I think we should remove it and use it in the follow up PRs we're planning?

talevy replied:
I will double check, but I believe it will be used by the index-name resolver when used for indexing. It is not used yet since we haven't implemented that yet

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -1040,6 +1059,34 @@ public MetaData build() {
+            AliasValidator.validateAliasWriteOnly(aliasAndIndexLookup);
+
+            // set any `is_write_index == null` to their appropriate boolean values
+            for (AliasOrIndex aliasOrIndex : aliasAndIndexLookup.values()) {
```
see other comment - this should all be done in buildAliasAndIndexLookup

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -1040,6 +1059,34 @@ public MetaData build() {
+
+
+            // build all concrete indices arrays:
+            // TODO: I think we can remove these arrays. it isn't worth the effort, for operations on all indices.
```
I believe this is about memory allocation - I think we can discuss this but in a different issue.

talevy replied:
I didn't write this

talevy commented on docs/reference/indices/aliases.asciidoc:
```diff
@@ -244,6 +244,94 @@ GET /alias2/_search?q=user:kimchy&routing=2,3
+// CONSOLE
+// TEST[s/^/PUT test\nPUT test2\n/]
+
+It is important to note that the order of these actions is important. The original write index
```
oh, woops. I thought I went through this. will update since it is left-over and currently wrong

talevy commented on server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetaData.java:
```diff
@@ -308,6 +337,8 @@ public static void toXContent(AliasMetaData aliasMetaData, XContentBuilder build
                 builder.field("search_routing", aliasMetaData.searchRouting());
             }
 
+            builder.field("is_write_index", Boolean.TRUE.equals(aliasMetaData.writeIndex()));
```
The aspect of Clint's original proposal that I liked was that we would always show this field and make it clear to the user which value we defaulted to. Having it be `null` feels confusing.

bleskes replied:
> make it clear to the user which value we defaulted to

That's my point - we don't know in this point. I agree that outputting `null` will be confusing hence my suggestion which will completely omit the field if not set.

talevy commented:
hey @bleskes, I've updated the build logic to behave more like a one-pass build of the aliasAndIndexLookup + manipulating the IndexMetaData of the cluster, itself.

one thing I am debugging is how this behavior is affecting snapshot restoration. Since one can restore and rename index names, the aliasmetadata now has multiple copies of `is_write_index: true`. So [restoration is failing](https://github.com/talevy/elasticsearch/blob/bdb0fb25555381299d80fed7efcbcab97a9d172f/server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java#L1540)

talevy commented on server/src/main/java/org/elasticsearch/cluster/metadata/AliasMetaData.java:
```diff
@@ -194,6 +210,11 @@ public AliasMetaData(StreamInput in) throws IOException {
+        }
     }
 
     public static Diff<AliasMetaData> readDiffFrom(StreamInput in) throws IOException {
```
I have a general question about cluster-state diffs, if you don't mind sharing. We seem to check stream-input version in the general Writeable.Reader, but I do not see any here. I'm not entirely sure why that is the case, shouldn't we also version-guard reading diffs across the wire?

talevy replied:
oh, nevermind. the diff is just a complete overwrite. not part-by-part, so the wire-version is covered

bleskes commented on server/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequestTests.java:
```diff
@@ -93,6 +93,7 @@ public void testToXContent() throws IOException {
         Alias alias = new Alias("test_alias");
         alias.routing("1");
         alias.filter("{\"term\":{\"year\":2016}}");
+        alias.writeIndex(true);
```
randomize?

bleskes commented on server/src/test/java/org/elasticsearch/cluster/metadata/AliasMetaDataTests.java:
```diff
@@ -95,6 +111,7 @@ private static AliasMetaData createTestItem() {
         if (randomBoolean()) {
             builder.filter("{\"term\":{\"year\":2016}}");
         }
+        builder.writeIndex(randomBoolean());
```
sometime make it null?

talevy replied:
I think it is ok to leave it as static since that is the nature of this test.

I believe testToAndFromXContent captures the random scenario, and so I will update it so
that randomAlias includes the write index randomization

bleskes commented on server/src/test/java/org/elasticsearch/cluster/metadata/AliasMetaDataTests.java:
```diff
@@ -54,6 +55,21 @@ public void testSerialization() throws IOException {
+                .filter(expectedInstance.filter())
+                .indexRouting(expectedInstance.indexRouting())
+                .searchRouting(expectedInstance.searchRouting())
+                .writeIndex(Boolean.FALSE)
```
why is that force false?

talevy replied:
just leaving the static nature of this test, will update

bleskes commented on server/src/test/java/org/elasticsearch/cluster/metadata/AliasValidatorTests.java:
```diff
@@ -44,4 +45,6 @@ public void testValidatesAliasNames() {
         // Doesn't throw an exception because we allow upper case alias names
         validator.validateAliasStandalone("CAT", null);
     }
+
```
left overs

bleskes commented on server/src/test/java/org/elasticsearch/cluster/metadata/IndexCreationTaskTests.java:
```diff
@@ -291,6 +296,32 @@ public void testValidateWaitForActiveShardsFailure() throws Exception {
     }
 
+    public void testWriteIndex() throws Exception {
+        Boolean writeIndex = randomBoolean();
```
random null?

bleskes commented on server/src/test/java/org/elasticsearch/cluster/metadata/ToAndFromJsonMetaDataTests.java:
```diff
@@ -111,7 +111,7 @@ public void testSimpleJsonFromAndTo() throws IOException {
                         .putMapping("mapping2", MAPPING_SOURCE2)
                         .putAlias(newAliasMetaDataBuilder("alias1").filter(ALIAS_FILTER1))
-                        .putAlias(newAliasMetaDataBuilder("alias2"))
+                        .putAlias(newAliasMetaDataBuilder("alias2").writeIndex(true))
```
some randomization comments

talevy replied:
as before, I chose to leave the static nature of this test. I can update here

bleskes commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataIndexAliasesServiceTests.java:
```diff
@@ -125,6 +128,63 @@ public void testRemoveIndexTwice() {
+            equalTo(after.metaData().index("test")));
+    }
+
+    public void testAddWriteOnlyWithExistingWriteIndex() {
```
I think we need more tests here - adding a write index to a group of non write indexes. removing an index from a default group of two indices (and seeing there is a write index)

talevy commented:
I've updated to add more tests and removed unused imports, ready for another review @bleskes. thanks!

talevy commented:
thanks! I will probably finalize the back-porting dance early next week in the hopes of not breaking a CI run on a Friday

talevy commented:
thanks @albertzaharovits! sorry for the noise
